File: code/com/leetcode/solutions/NO_0016_最接近的三数之和/Solution.js
/**
 * 思路：参照三数之和
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function (nums, target) {
    nums.sort((a, b) => a - b);
    const len = nums.length, e = len - 2, s = len - 1;
    let ans = nums[0] + nums[1] + nums[2] - target, sum, i = 0;

    for (let j, k; i < e; i++) {
        j = i + 1;
        k = s;
        while (j < k) {
            sum = nums[i] + nums[j] + nums[k] - target;
            if (sum > 0) k--;
            else if (sum < 0) j++;
            else {
                ans = 0;
                break;
            }
            if (Math.abs(sum) < Math.abs(ans)) {
                ans = sum;
            }
        }
        if (ans === 0) break;
    }
    return ans + target;
};

File: code/com/leetcode/solutions/NO_0016_最接近的三数之和/Solution.java
package com.leetcode.solutions.NO_0016_最接近的三数之和;

/**
 * Created by gang.xu01@hand-china.com on 2020/7/6
 *
 * @author gang.xu01@hand-china.com
 */
public class Solution {

    public int threeSumClosest(int[] nums, int target) {
        // 定义三个指针
        int count = nums[0] + nums[1] + nums[2];
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                for (int k = j + 1; k < nums.length; k++) {
                    count = Math.abs(target - nums[i] - nums[j] - nums[k]) > Math.abs(target - count)
                            ? count : nums[i] + nums[j] + nums[k];
                }
            }
        }
        return count;
    }

}
